[Workers] WASM Rust Guide (#8539)
cloudflare/cloudflare-docs: content/workers/platform/web-assembly/_index.md

---
pcx_content_type: concept

> title: WebAssembly (WASM)

Reviewer: ```suggestion
title: WebAssembly (Wasm)
```

weight: 2
---

> # WebAssembly (WASM)

Reviewer: ```suggestion
# WebAssembly (Wasm)
```

> [WebAssembly](https://webassembly.org/) (abbreviated Wasm) is a binary instruction format for a stack-based virtual machine.

Reviewer: Literally true and probably worth including, but I think most people arriving at these docs will resonate with something more akin to:

```suggestion
[WebAssembly](https://webassembly.org/) (abbreviated Wasm) allows you to compile languages like Rust, Go, or C a binary format that can run in a wide variety of environments, including [web browsers](https://developer.mozilla.org/en-US/docs/WebAssembly#browser_compatibility), Cloudflare Workers, and other WebAssembly runtimes.

On Workers, you can use WebAssembly to:

- Execute code written in a language other than JavaScript, via `WebAssembly.instantiate()` <link to this section>
- Write an entire Cloudflare Worker in Rust, using bindings that make Workers' JavaScript APIs available directly from Rust <link to section>
```

(this last section too verbose, can think on how to clean up)

> Many [languages](http://localhost:5174/workers/platform/languages/#wasm-supported) can be compiled to Wasm and run on Workers.

Reviewer: ```suggestion
Many [languages](/workers/platform/languages/#wasm-supported) can be compiled to Wasm and run on Workers.
```

Reviewer: Maybe omit since this list of languages doesn't represent the full set of what is possible, only what we have links for? And edit above names a few?

> Guides are available for the following languages:

Reviewer: ```suggestion
Most programming languages can be compiled to Wasm, though support varies across languages and compilers. Guides are available for the following:
```

{{<directory-listing>}}

## Supported Proposals

> WebAssembly includes [many proposed APIs](https://webassembly.org/roadmap/) which are in various stages of development. In general, Workers supports "standardized" features which have been implemented by V8, except where the feature may pose a security risk.

Reviewer: ```suggestion
WebAssembly is a rapidly evolving set of standards, with [many proposed APIs](https://webassembly.org/roadmap/) in various stages of development. In general, Workers supports the same set of features that are available in Google Chrome
```

If ^ is a directionally accurate statement, maybe clearer reference than V8?

### SIMD

SIMD is supported on Workers. For more information on using SIMD in WebAssembly, see [Fast, parallel applications with WebAssembly SIMD](https://v8.dev/features/simd).

### Threading

Threading is not possible in Workers because each Worker runs in a single thread to mitigate timing attacks, and the [Web Worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) API is not supported.

## Binary Size

> WebAssembly tends to produce larger Workers bundles than JavaScript Workers. Cloudflare limits the size of Workers bundles for performance reasons. This limit will increase over time, however larger bundle size may still impact script startup time. You should

Reviewer: ```suggestion
Compiling to WebAssembly often requires including additional runtime dependencies. As a result, Workers that use WebAssembly are typically larger than an equivalent Worker written in JavaScript. The larger your Worker is, the longer it may take your Worker to [start](https://developers.cloudflare.com/workers/platform/limits/#worker-startup-time).
```

> use tools like [`wasm-opt`](https://github.com/brson/wasm-opt-rs) to optimize the WASM binary, and limit the number of dependencies

Reviewer: ```suggestion
We recommend using tools like [`wasm-opt`](https://github.com/brson/wasm-opt-rs) to optimize the size of your Wasm binary, and limit the number of dependencies included in your Worker.
```

> included in your Worker.

Reviewer: ```suggestion
```

> ## WASI

Reviewer: TODO: Explain what WASI is @irvinebroque

> WASI support is experimental, with only some syscalls implemented. See our open source implementation [here](https://github.com/cloudflare/workers-wasi), and [blog post](https://blog.cloudflare.com/announcing-wasi-on-workers/) demonstrating its use. 

Reviewer: TODO for @irvinebroque — make these a table

---
pcx_content_type: concept
title: WebAssembly (Wasm)
---

# WebAssembly (Wasm)

[WebAssembly](https://webassembly.org/) (abbreviated Wasm) allows you to compile languages like Rust, Go, or C to a binary format that can run in a wide variety of environments, including [web browsers](https://developer.mozilla.org/en-US/docs/WebAssembly#browser_compatibility), Cloudflare Workers, and other WebAssembly runtimes.

On Workers, you can use WebAssembly to:
- Execute code written in a language other than JavaScript, via `WebAssembly.instantiate()`
- Write an entire Cloudflare Worker in Rust, using bindings that make Workers' JavaScript APIs available directly from your Rust code.

> Most programming languages can be compiled to Wasm, although support varies across languages and compilers. Guides are available for the following:

Reviewer 2: I don't know if I'd go as far as most programming languages, I think many is a much more appropriate term here.
